### PR TITLE
Optimal huff depth

### PR DESCRIPTION
**Pull request by daniellerozenblit**

### CODE REVIEW

**Comment by daniellerozenblit** [comment]
## TLDR
This PR modifies `HUF_optimalTableLog` so that it behaves differently for high and low compression levels. Previously, we used the same heuristic to find the table log for all compression levels. This PR introduces a change that, for high compression levels, tests all valid table depths and chooses the optimal based on minimizing encoded + header size. This allows us to find additional compression ratio gains in the entropy stage for high compression levels.

The threshold to find the optimal table log is compression level 13 (or strategy ZSTD_btlazy2). Any compression level below this will continue to use the previous heuristic.

## Benchmarking
I benchmarked on an Intel(R) Xeon(R) Gold 6138 CPU @ 2.00GHz machine with core isolation and turbo disabled. I measured compression time and compression ratio for silesia.tar (212M), compiling with clang15. I ran each scenario 5 times and chose the minimum value. 

### Compression Ratio
| Compression Level | Dev Size (B) | Test Size (B) | Size Delta (B)|
|-------------------|--------------|---------------|------------------|
| 3                 | 66508266 | 66507319| -947 |
| 6                 | 61468353 | 61466030 | -2323 |
| 9                 | 59288102 | 59286414 | -1688 |
| 12                | 58180337 | 58178678 | -1659 |
| 15                | 57159315 | 57157583 | -1732 |
| 18                | 53452679 |53447936| -4743|
| 19                | 53007368 |53002324| -5044|

### Compression Speed
| Compression Level | Dev Speed (Mb/s) | Test Speed (Mb/s) | Speed Delta (%) |
|-------------------|------------------|-------------------|-----------------|
| 3                 | 130.8           | 130.4             |-0.306|
| 6                 | 52.7           | 52.0             |-1.33|
| 9                 | 33.7           | 32.6             | -3.26 |
| 12                | 17.4           | 17.3             | −0.575|
| 15                | 5.31           | 5.29             |−0.377|
| 18                | 2.28            | 2.26| -0.877| 
| 19                | 1.85            | 1.83| −1.081 |
